Feature/geometric stream extraction #128

richpsharp opened the pull request:
This PR adds new new functions to `pygeoprocessing.routing` and closes issue #128 

* `extract_strahler_streams_d8` - creates geometric line segments corresponding to streams in a D8 flow raster and automatically determines a reasonable flow accumulation value per river with a statistical test for expected drop distance between streams.
 * `calculate_subwatershed_boundary` - uses the stream layer from above to calculate subwatersheds using a perimeter tracing technique.

Typical output from `extract_strahler_streams_d8` may look like this
![image](https://user-images.githubusercontent.com/7773490/103606569-99c3b680-4ecb-11eb-948d-fc2084e8e203.png)

And the subwatershed calculation would look like this: 
![image](https://user-images.githubusercontent.com/7773490/103606724-fde67a80-4ecb-11eb-9748-33e2600451f1.png)

This PR includes new code for both functions, supporting helper functions, a large full integration test, and a small history update.

Runtime seems reasonable with performance of the stream extraction taking approximately 2 minutes for a 6k+ X 5k+ raster and 10ish seconds for the subwatershed calculation.

richpsharp committed:
re #128 adding a function header for extracting the strahler stream layer

richpsharp committed:
driving a loop to make the base geometry

richpsharp committed:
filling out first step upstream

richpsharp committed:
making a 1D array for the backflow and just using flat indexing... i was getting a memory error or something before

richpsharp committed:
re #128 adding per bifurcation point

richpsharp committed:
re #128 created stream geometry extraction layer

richpsharp committed:
re #128 seeded order as 1

richpsharp committed:
re #128 got geometric streams and working on mahler order

richpsharp committed:
Merge branch 'main' into feature/geometric-stream-extraction-128

richpsharp committed:
re #128 refactoring to use the feature FID all the way through instead of an arbitrary one

richpsharp committed:
re #128 complete geometric stream extraction

richpsharp committed:
re #128 set bit to determine when it's an outlet

richpsharp committed:
re #128 adding a backtrace to uniquely ID the streams

richpsharp committed:
do the test on the right stream

richpsharp committed:
re #128 using integer lists to record source point

richpsharp committed:
re #128 adding in code to adjust streams as needed

richpsharp committed:
re #128 middle of refactor

richpsharp committed:
re #128 returning raster space coord rather than projected coord

richpsharp committed:
simplify stream ending

richpsharp committed:
re #128 fixing issue where streams weren't getting updated

richpsharp committed:
re #128 fixing issue with setting downstream order

richpsharp committed:
re #128 fixed issue with missing 1 pixel streams

richpsharp committed:
re #128 created discovery and finish time rasters

richpsharp committed:
re #128 adding a first step for boundary walk

richpsharp committed:
re #128 fixing bad finish time that was off by 1

richpsharp committed:
re #128 adding the step to draw the edges on the rest of the raster

richpsharp committed:
re #128 fixed an issue that was preventing the route from completing.

richpsharp committed:
re #128 walking the edge correctly

richpsharp committed:
re #128 fixed issue where some starts were offset the wrong direction in 5

richpsharp committed:
re #128 fixed issue with watershed satarting off on wrong pixel

richpsharp committed:
re #128 working to do leaf first iteration

richpsharp committed:
re #128 fixed issue with watershed intersecting on way back up

richpsharp committed:
backputting the order 1 stream that is an outlet

richpsharp committed:
re #128 fix issue where first step on watershed might not indicate a boundary

richpsharp committed:
re #128 fixed many bugs, working on few corner cases

richpsharp committed:
re #128 fixed many bugs, working on few corner cases

richpsharp committed:
re #128 fixing issue where order 1 streams were getting root filled before junctioned higher order streams

richpsharp committed:
re #128 fixed issue with outlets not triggering upstream junction watershed

richpsharp committed:
re #128 cleaning up logging messages

richpsharp committed:
re #128 docstring typo

richpsharp committed:
re #128 adding a test case for strahler streams

richpsharp committed:
re #128 making a test that covers the new stream/watershed pipeline

richpsharp committed:
fixes #128 updating history

richpsharp committed:
re #128 naming the function subwatershed instead of watershed

richpsharp committed:
Merge branch 'main' into feature/geometric-stream-extraction-128

richpsharp committed:
re #128 adding streams and subwatershed to __init__ export

richpsharp committed:
Merge branch 'main' into feature/geometric-stream-extraction-128

phargogh commented on HISTORY.rst:
```diff
@@ -25,6 +25,13 @@ Unreleased Changes
+  ``pygeoprocessing.extract_strahler_streams_d8``. Creates segmented
+  geometric lines which correspond to streams on the landscape.
+* Added a D8 subwatershed calculation function at
+  ``pygeoprocessing.calculate_watershed_boundary``. Creates subwatersheds
```
Wouldn't this be `calculate_subwatershed_boundary` rather than `calculate_watershed_boundary`?

richpsharp replied:
It should be! And I just noticed i had it in the wrong part of the history file -- 2.1.2 release, not 'unreleased'. Fixed both in eaedf452dbec0be690ca17d88e81af7928429339

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -77,6 +88,9 @@ cdef int* NEIGHBOR_OFFSET_ARRAY = [
     1, 1  # 7
     ]
 
+cdef int *D8_XOFFSET = [1, 1, 0, -1, -1, -1, 0, 1]
+cdef int *D8_YOFFSET = [0, -1, -1, -1, 0, +1, +1, +1]
```
I realize these and `NEIGHBOR_OFFSET_ARRAY` are basically the same thing just in ever-so-slightly different presentations, but would it be worth consolidating them or deriving one from the other to reduce duplication?

richpsharp replied:
Yep I like it. Done in f1812913d3416a5b8269d5e2cdabeec841342b47

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1287 @@ def _is_raster_path_band_formatted(raster_path_band):
+            flow direction raster.
+        flow_accum_raster_path_band (tuple): a path/band representing the D8
+            flow accumulation raster represented by
+            `flow_dir_d8_raster_path_band`.
```
Double-backticks for RST?

richpsharp replied:
Fixed in 0d03242574f2801b124cab5a526dec4897aa8c2f

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1287 @@ def _is_raster_path_band_formatted(raster_path_band):
+    at least the fields:
+        * "order" (int): an integer representing the stream order
+        * "river_id" (int): unique ID used by all stream segements that
+            connect to the same outlet.
```
I can't find the RTD build for this PR to verify, I think we might need a blank line above the bulleted list for it to be interpreted as a list.

richpsharp replied:
Fixed in 0d0081f

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1287 @@ def _is_raster_path_band_formatted(raster_path_band):
+    at least the fields:
+        * "order" (int): an integer representing the stream order
+        * "river_id" (int): unique ID used by all stream segements that
+            connect to the same outlet.
```
There are a bunch of other fields that are also written to the vector, would it be worth documenting them here as well?

richpsharp replied:
Weeeelll... I did say "at least the following fields" because in a way they're useful for something else, but I guess that should be documented if it really is useful huh?

richpsharp replied:
Okay i documented all the features in 0903796a7b389b1123f4176f62774c1ca9f82f30 except "d8_flow_dir" which was just an intermediate field anyway that I delete now before the function is over.

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1287 @@ def _is_raster_path_band_formatted(raster_path_band):
+    stream_vector = gpkg_driver.Create(
+        target_stream_vector_path, 0, 0, 0, gdal.GDT_Unknown)
+    stream_layer = stream_vector.CreateLayer(
+        'stream', flow_dir_srs, ogr.wkbLineString)
```
With watershed delineation, Stacie ended up requesting that we add the file suffix to the layer name as well, since it's the layer name that is actually shown in the Arc and QGIS.  With watershed delineation, the solution we settled on was to [allow a user-defined layer name](https://github.com/natcap/pygeoprocessing/blob/main/src/pygeoprocessing/routing/watershed.pyx#L628).  Maybe this would be useful here as well?

richpsharp replied:
oh sure, in fact i think I've always just used the base filename which is what i'd prefer

richpsharp replied:
done in c46e5741c1948a29fa0da521f6dafb35471ad4fe

phargogh replied:
Oh yeah, that's even better.  Thanks!

One thing I noticed, though is that c46e5741c1948a29fa0da521f6dafb35471ad4fe also includes the new function from #148, which doesn't look intentional!

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1287 @@ def _is_raster_path_band_formatted(raster_path_band):
+        1, 0, 0, 0, 0, 0, 0, 0,
+        0, 1, 0, 0, 0, 0, 0, 0,
+        0, 0, 1, 0, 0, 0, 0, 0,
+        0, 0, 0, 1, 0, 0, 0, 0]
```
Neat adjacency matrix!  I see this same matrix used in 3 places throughout the file ... would it be worth abstracting out to a module-level definition?

And on a related note, it also looks like we have a similar backflow array represented in `D8_REVERSE_DIRECTION ` at the module level.  I realize that their representations are slightly different, but would it be worth the effort to consolidate the approaches?

richpsharp replied:
Ugh you're right, `D8_REVERSE_DIRECTION` is superior and I dumped d8_backflow in lieu of that in fb837a1.

richpsharp committed:
merge

richpsharp committed:
n left turns to guard against loop

richpsharp committed:
re #128 making abs and rel values explict when testing _is_close, fixing issue of is close not being that close when checking loop around, and cleaning up error messages

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -409,9 +439,21 @@ cdef class _ManagedRaster:
-            raster = gdal.OpenEx(
-                self.raster_path, gdal.GA_Update | gdal.OF_RASTER)
-            raster_band = raster.GetRasterBand(self.band_id)
+            n_attempts = 5
+            while True:
+                raster = gdal.OpenEx(
+                    self.raster_path, gdal.GA_Update | gdal.OF_RASTER)
+                if raster is None:
+                    if n_attempts == 0:
+                        raise RuntimeError(
+                            f'could not open {self.raster_path} for writing')
+                    LOGGER.warning(
+                        f'opening {self.raster_path} resulted in null, '
+                        f'trying {n_attempts} more times.')
+                    n_attempts -= 1
+                    time.sleep(0.5)
+                raster_band = raster.GetRasterBand(self.band_id)
+                break
```
Hey, great idea!  Was this something that same up strictly in the context of debugging this new routing functionality, or should it be replicated elsewhere where we have `ManagedRaster` classes defined?

richpsharp replied:
It came up when testing the subwatershed routing... that's not to say i wouldn't want it everywhere else but it was specifically happening during this development. What I'd *really* like is to avoid duplicating this class in every Cython module, but every time I try to abstract this into a Cython class things start to break... I think bevcause of the linkage to the LRUCache.h file. I think we've both attempted it over the years. Can I punt on duplicating this code everywhere else and if it's an issue in the future we try really hard to get a ManagedRaster abstract component we can use in all the pygeoprocessing cython modules?

phargogh replied:
Oh for sure!  We've definitely both tried to abstract this out, but it became clear that it's going to take a C++ extension to really do it right with templating.  So yes, let's punt.

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+            point_discovery <= finish)
+
+
+def _calculate_stream_geometry(
```
Would it be worth `cdef`ing this function?  Most of the parameter types are already defined, so I wonder if that was the original intent here.  Based on where it's being used in this file, I'd guess it wouldn't be a critical speedup.

richpsharp replied:
Sure! changed in f97cfcd. I think I left it as a def because it returns a geometry object so there's overhead on that but it doesn't hurt anything. Mostly this is in a function because it's a complicated bit of code I needed to use a few different places and since we were already in Cython I static typed everything that made sense to.

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+        1, 0, 0, 0, 0, 0, 0, 0,
+        0, 1, 0, 0, 0, 0, 0, 0,
+        0, 0, 1, 0, 0, 0, 0, 0,
+        0, 0, 0, 1, 0, 0, 0, 0]
```
Here's another case of the D8 backflow adjacency matrix that could be consolidated to the top of the module.

richpsharp replied:
Fixed in previous commit

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+        Georeferenced linestring connecting x/y to upper point where upper
+            point's threshold is the last point where its flow accum value
+            is >= `flow_accum_threshold`. Returns None if x/y is below
+            flow accum threshold.
```
Looks like the return value has changed since this returns item was written!  Could be worth updating.

richpsharp replied:
Update the return docstring in 8d6eea1

phargogh replied:
Nice!  I like the RST formatting, as well :)

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+    upstream_id_list = []
+    # anchor the line at the downstream end
+    stream_line = ogr.Geometry(ogr.wkbLineString)
+    x_p, y_p = gdal.ApplyGeoTransform(geotransform, x_l+0.5, y_l+0.5)
```
TIL about `gdal.ApplyGeoTransform`.  Neat!  I bet there are more than a few places where we could be using this function in InVEST and pygeoprocessing.

richpsharp replied:
I use it all the time. Also useful in combination with `gdal.InvGeoTransform` if you want to turn projected coordinates into raster coordinates.

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+        1, 0, 0, 0, 0, 0, 0, 0,
+        0, 1, 0, 0, 0, 0, 0, 0,
+        0, 0, 1, 0, 0, 0, 0, 0,
+        0, 0, 0, 1, 0, 0, 0, 0]
```
And here's another backflow adjacency matrix.

richpsharp replied:
fixed fixed!

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+
+    Returns:
+        linestring of projected points indicating watershed boundary.
+    """
```
This docstring looks to be a bit out of date relative to the function signature and what the function appears to be returning.

richpsharp replied:
Fixed in e057a8b3d95b96091a3189c4b41d1c8fae8399ba

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+    """
+    workspace_dir = tempfile.mkdtemp(
+        dir=os.path.join(
+            os.path.dirname(target_watershed_boundary_vector_path)))
```
Could I recommend using one of the prefix/suffix naming arguments to clarify what the tmpdir is being used for?  I've personally found it useful when there are a bunch of temporary directories lying around.

richpsharp replied:
Sure, added in e1afc69

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+                boundary_list.append((x_l, y_l))
+
+        # use to keep track of how many turns around a pixel, used if
+        # a degerate
```
"Degenerate"?

richpsharp replied:
hilarious and also unnecessary since that bit of code wasn't needed at all. i removed the command and all the  "n_turns" bits too in 28fbda7

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+            # step the edge then determine the projected coordinates
+            x_f += D8_XOFFSET[edge_dir]
+            y_f += D8_YOFFSET[edge_dir]
+            # equivalent to gdal.ApplyGeoTransform(geotransform, x_f, y_f):
```
Are the `gdal.ApplyGeoTransform` equivalents below just to eliminate the python function call overhead?

richpsharp replied:
Yes and i realize there are other python calls in there, but they're so few i figure if i can cut one out of 4 that's 1/4 speedup? i added a comment in there about that too in 361d9e1

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+
+            if n_left_turns == 4:
+                LOGGER.info(
+                    '4 left turns around the same pixel so terminating')
```
Would it be useful to log some other information about this feature, such as the FID of the watershed?

richpsharp replied:
That line is no longer in there because i no longer explicitly check for 4 turns as in 28fbda7 above

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+        boundary_list (list): this list is appended to for new pixels that
+            should be neighbors in the fill.
+
+    Return:
+        None.
```
There's no return value, but there is a side effect.  This is a CDEF'd function so probably no one will actually see this unless they're reading the source, but should the side effect be recorded here?

richpsharp replied:
I think I do document that side effect in the 'args' as "boundary_list (list): this list is appended to for new pixels that should be neighbors in the fill." Is that okay?

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+    watershed_vector = None
+    discovery_managed_raster.close()
+    finish_managed_raster.close()
+    #shutil.rmtree(workspace_dir)
```
Lingering deprecated code?

richpsharp replied:
So weird, that line is not in my working file, but otherwise yes, I would have removed it.

phargogh reviewed (changes requested):
Thanks, Rich!  I had a few comments, mostly pretty minor, and the algorithms themselves look really interesting.  Let me know what you think!

richpsharp committed:
re #128 indicating subwatershed fucntion name and putting it in the right part of the history file

richpsharp committed:
re #128 consolidating NEIGHBOR_OFFSET_ARRAY to the DX/DY_OFFSET_ARRAY

richpsharp committed:
re #128 double backticks

richpsharp committed:
re #128 adding blank line so it's a bulleted list

richpsharp committed:
re #148 removing upstream_d8_dir and documenting all the other fields added to the stream layer

richpsharp committed:
re #128 removing unused field for d8 dir and spelling the d8_xoffset arrays right

richpsharp committed:
#128 removing n turns around a pixel because we test for that directly

richpsharp committed:
re #128 adding to eliminate python function call overhead for clarity

richpsharp committed:
merge

richpsharp commented:
Hi @phargogh ! I just addressed all your comments which were all great. My favorite was consolidating those two different ways to determine x/y offsets for neighbors.

richpsharp committed:
re #128 naming the layer after the basename of the file

richpsharp committed:
re #128 consolidating d8_backflow to a global

richpsharp committed:
re #128 for some reason shared memory keeps getting obliterated from multiprocessing

richpsharp committed:
re #148 encountered a flaky test that would crash because the workspace was unable to be removed due to a process still using the raster. the __swig_destroy__ in that case cleans up the objhect

richpsharp committed:
re #128 turning def into cdef

richpsharp committed:
re #128 fixing the return docstring to indicat it's a tuple

richpsharp committed:
re #128 updating docstring with simpler prototype

richpsharp committed:
re #128 using a prefix

richpsharp commented:
Okay @phargogh , I know earlier I said I addressed all your comments but github had compressed the conversation and I didn't notice at first that 10 more of your comments were hidden. So I did one more pass and you'll notice one other change I made was to ADD a `__swig_destroy__` back into pygeoprocessing.rasterize. For some reason a test case started to fail where I cause the callback handler to trigger a divide by zero exception, but then the test case couldn't clean up the working directory because the raster was still in use. I tried a LOT of things and the only thing that worked in the end was a specific `__swig_destroy__` call on the raster when that particular exception is triggered. So sorry that's tossed in here too.

richpsharp committed:
re #128 replacging d8_backflow with d8_reverse direction

phargogh commented on src/pygeoprocessing/routing/routing.pyx:
```diff
@@ -2948,6 +2983,114 @@ def extract_streams_mfd(
     LOGGER.info('100.0% complete')
 
 
+def detect_outlets(d8_flow_dir_raster_path_band, target_outlet_vector_path):
```
I think this somehow made its way into this PR, though it's also under review in #148 ... I think that wasn't the intent?

richpsharp replied:
Sure wasn't the intent and honestly not sure how it got in there. And double honestly I'm still not comfortable with git's branching, I can't seem to unroll it/detect what happened as easily as mercurial. Anyway, I've removed `detect_outlets` in 593996ac11670eb709a1327fe98bf961e4369f7f.

phargogh reviewed (changes requested):
Thanks, Rich!  I just had one last comment about a function being reviewed elsewhere that seems to have made its way into this file.  But other than that, I think this is looking good.

richpsharp committed:
re #128 outlet detection accidentally in there

richpsharp commented:
> Thanks, Rich! I just had one last comment about a function being reviewed elsewhere that seems to have made its way into this file. But other than that, I think this is looking good.

thanks @phargogh, I removed the function, pushed, tests pass etc.

phargogh reviewed (approved):
Merging!
